assembly/src/parsers/ast/mod.rs

```diff
@@ -0,0 +1,339 @@
+
+    // parse the sequence of nodes and add each node to the list
+    let mut end_of_nodes = false;
+    let beginning_node_count = body.len();
+    while !end_of_nodes {
+        let node_count = body.len();
+        context.parse_body(&mut tokens, &mut body)?;
+        end_of_nodes = body.len() == node_count;
+    }
+
+    // make sure at least one block has been read
+    if body.len() == beginning_node_count {
+        let start_op = tokens.read_at(start_pos).expect("no start token");
+        return Err(AssemblyError::empty_block(start_op));
+    }
+
+    // consume the 'end' token
+    match tokens.read() {
+        None => Err(AssemblyError::unmatched_begin(
+            tokens.read_at(program_start).expect("no begin token"),
+        )),
+        Some(token) => match token.parts()[0] {
+            Token::END => token.validate_end(),
+            Token::ELSE => Err(AssemblyError::dangling_else(token)),
+            _ => Err(AssemblyError::unmatched_begin(
+                tokens.read_at(program_start).expect("no begin token"),
+            )),
+        },
+    }?;
+    tokens.advance();
+
+    // make sure there are no instructions after the end
+    if let Some(token) = tokens.read() {
+        return Err(AssemblyError::dangling_ops_after_program(token));
+    }
+
+    let program = ProgramAst {
+        body,
+        procedures: context.procedures,
+    };
+
+    Ok(program)
+}
+
+pub fn parse_module(source: &str) -> Result<ModuleAst, AssemblyError> {
+    let mut tokens = TokenStream::new(source)?;
+
+    let imports = parse_imports(&mut tokens)?;
+    let mut context = ParserContext {
+        procedures: BTreeMap::new(),
+        imports,
+    };
+    context.parse_procedures(&mut tokens, true)?;
+
+    // make sure program body is absent and no more instructions.
+    if tokens.read().is_some() {
+        return Err(AssemblyError::unexpected_eof(tokens.pos()));
+    }
+
+    let module = ModuleAst {
+        procedures: context.procedures,
+    };
+
+    Ok(module)
+}
+
+fn parse_op_token(op: &Token) -> Result<Node, AssemblyError> {
+    // based on the instruction, invoke the correct parser for the operation
+    let node = match op.parts()[0] {
+        // ----- field operations -----------------------------------------------------------------
+        "assert" => Node::Instruction(Instruction::Assert),
+        "assert_eq" => Node::Instruction(Instruction::AssertEq),
+        "assertz" => Node::Instruction(Instruction::Assertz),
+
+        "add" => Node::Instruction(Instruction::Add),
+        "sub" => Node::Instruction(Instruction::Sub),
+        "mul" => Node::Instruction(Instruction::Mul),
+        "div" => Node::Instruction(Instruction::Div),
+        "neg" => Node::Instruction(Instruction::Neg),
+        "inv" => Node::Instruction(Instruction::Inv),
+
+        "checked_pow2" => Node::Instruction(Instruction::Pow2Checked),
+        "unchecked_pow2" => Node::Instruction(Instruction::Pow2Unchecked),
```

Comment:
This has been changed now: we have only `pow2` (no checked vs. unchecked), and we also now have an additional `exp` instruction.

Reply:
fixed